File: enrich_shownotes.py

from bs4 import BeautifulSoup
import re
import os

HTML_FILE = 'shownotes/GenZ_AILiteracy/index.html'

def clean_slug(url):
    # Remove trailing slash
    if url.endswith('/'):
        url = url[:-1]
    
    # Get last part
    parts = url.split('/')
    slug = parts[-1]
    
    # If slug is query string (e.g. ?v=...), try previous part?
    # actually for standard URLs usually last part is good.
    # for youtube, slug is 'watch?v=...' or similar.
    
    # Clean up
    # Replace - _ + with space
    text = re.sub(r'[-_+]', ' ', slug)
    
    # Remove file extension if present (e.g. .html, .php)
    if '.' in text:
        text = text.rsplit('.', 1)[0]
        
    # Title Case
    # Split by camelCase if needed? simpler is just built-in title()
    return text.title()

def enrich_html():
    if not os.path.exists(HTML_FILE):
        print(f"File not found: {HTML_FILE}")
        return

    with open(HTML_FILE, 'r', encoding='utf-8') as f:
        soup = BeautifulSoup(f, 'html.parser')

    # Find all 'a' tags
    links = soup.find_all('a')
    
    count = 0
    for link in links:
        # Check text content. 
        # For video cards, the text is in a span
        span = link.find('span')
        if span:
            current_text = span.get_text().strip()
        else:
            current_text = link.get_text().strip()
            
        if current_text.lower() == 'link':
            href = link.get('href', '')
            if not href:
                continue
                
            new_text = ""
            
            # Logic for YouTube
            if 'youtube.com' in href or 'youtu.be' in href:
                # Look for preceding paragraph
                prev = link.find_previous_sibling()
                # Skip string nodes
                while prev and isinstance(prev, str):
                     prev = prev.find_previous_sibling()
                
                if prev and prev.name == 'p':
                    # Use paragraph text
                    new_text = prev.get_text().strip()
                    # Clean up leading '* ' if present
                    if new_text.startswith('* '):
                        new_text = new_text[2:]
                else:
                    new_text = "Watch Video"
            else:
                # Logic for other URLs: use slug
                new_text = clean_slug(href)
                # Fallback if slug is empty or weird
                if len(new_text) < 3:
                     # Try domain or second to last part?
                     new_text = "Read Article"

            # Check for bad slugs like "Index.Php" or "Watch"
            if new_text.lower() in ['index', 'watch', 'default']:
                 new_text = "Read Article"

            # Apply new text
            if span:
                span.string = new_text
                # Also update img alt if generic
                img = link.find('img')
                if img and img.get('alt') == 'Link':
                    img['alt'] = new_text
            else:
                link.string = new_text
                
            print(f"Updated: {href[:40]}... -> {new_text}")
            count += 1
            
    with open(HTML_FILE, 'w', encoding='utf-8') as f:
        f.write(str(soup.prettify()))
        
    print(f"Total links updated: {count}")

if __name__ == "__main__":
    enrich_html()

File: enrich_shownotes.js

import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

const HTML_FILE = path.resolve('shownotes/GenZ_AILiteracy/index.html');

async function main() {
    console.log(`Processing ${HTML_FILE}...`);

    const browser = await puppeteer.launch();
    const page = await browser.newPage();

    // Load the local HTML file
    const fileUrl = 'file://' + HTML_FILE.replace(/\\/g, '/');
    await page.goto(fileUrl, { waitUntil: 'domcontentloaded' });

    // 1. Add IDs to Part Headers for TOC
    await page.evaluate(() => {
        const headers = Array.from(document.querySelectorAll('h2.section-title'));
        headers.forEach((h2) => {
            const text = h2.textContent.trim().toLowerCase();
            if (text.includes('part 1') || text.includes('part i:')) h2.id = 'part1';
            else if (text.includes('part ii') || text.includes('part 2')) h2.id = 'part2';
            else if (text.includes('part iii') || text.includes('part 3')) h2.id = 'part3';
            else if (text.includes('part iv') || text.includes('part 4')) h2.id = 'part4';
            else if (text.includes('part v') || text.includes('part 5')) h2.id = 'part5';
            else if (text.includes('part vi') || text.includes('part 6')) h2.id = 'part6';
            else if (text.includes('part vii') || text.includes('part 7')) h2.id = 'part7';
        });
    });

    // 2. Identify generic links
    const linksToEnrich = await page.evaluate(() => {
        const anchors = Array.from(document.querySelectorAll('a'));
        return anchors.map((a, index) => {
            const text = a.textContent.trim();
            const href = a.href;

            // Check if text is generic "Link" or similar
            const isGeneric = text.toLowerCase() === 'link' || text.toLowerCase() === 'watch video' || text === href;

            if (isGeneric && href && !href.startsWith('mailto:') && !href.startsWith('file:')) {
                // Return index to identify element later
                return { index, href, isYoutube: (href.includes('youtube.com') || href.includes('youtu.be')) };
            }
            return null;
        }).filter(item => item !== null);
    });

    console.log(`Found ${linksToEnrich.length} links to enrich.`);

    // 3. Process links
    for (const link of linksToEnrich) {
        let newTitle = '';

        if (link.isYoutube) {
            try {
                // Fetch YouTube Title
                console.log(`Fetching YouTube title for: ${link.href}`);
                const ytPage = await browser.newPage();
                await ytPage.goto(link.href, { waitUntil: 'domcontentloaded' });
                newTitle = await ytPage.title();
                await ytPage.close();

                // Clean up title (remove "- YouTube")
                newTitle = newTitle.replace(/- YouTube$/, '').trim();
                console.log(`  -> Title: ${newTitle}`);
            } catch (e) {
                console.error(`  Failed to fetch title for ${link.href}: ${e.message}`);
                newTitle = 'Watch Video';
            }
        } else {
            // Slug to Title conversion
            try {
                const urlObj = new URL(link.href);
                let slug = urlObj.pathname.split('/').filter(p => p).pop();
                if (!slug) slug = urlObj.hostname;

                // Remove extension
                if (slug.includes('.')) slug = slug.split('.').slice(0, -1).join('.');

                // Simple cleanup
                newTitle = slug.replace(/[-_]/g, ' ');
                // Title case
                newTitle = newTitle.replace(/\b\w/g, c => c.toUpperCase());

                if (newTitle.length < 3) newTitle = "Read Article";

                console.log(`  Converted slug for ${link.href} -> ${newTitle}`);
            } catch (e) {
                newTitle = 'Read Article';
            }
        }

        // Update the link in the page context
        if (newTitle) {
            await page.evaluate((index, title) => {
                const anchors = Array.from(document.querySelectorAll('a'));
                const a = anchors[index];
                if (a) {
                    // Update text content of span if exists (for video cards)
                    const span = a.querySelector('span');
                    if (span) span.textContent = title;
                    else a.textContent = title;

                    // Update alt text of image if generic
                    const img = a.querySelector('img');
                    if (img && (img.alt === 'Link' || img.alt === '')) img.alt = title;
                }
            }, link.index, newTitle);
        }
    }

    // 4. Save modified HTML
    const content = await page.content();
    fs.writeFileSync(HTML_FILE, content);
    console.log('Saved updated HTML.');

    await browser.close();
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
